rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -449,22 +467,64 @@ extern "C" rmw_ret_t rmw_set_log_severity(rmw_log_severity_t severity)
   return RMW_RET_OK;
 }
 
+static void dds_listener_callback(dds_entity_t entity, void * arg)
+{
+  // Currently we don't use the entity which calls this callback.
+  // We need the dds_entity_t arg just to comply with the function pointer
+  // signature of 'dds_on_data_on_readers_fn' defined on dds_public_listener.h
+  (void)entity;
+
+  // Cast the listener callback arg to user callback data.
+  struct user_callback_data_t * data;
+  data = static_cast<user_callback_data_t *>(arg);
+
+  // Call the user callback
+  data->callback(
+    data->user_data,
+    {data->entity_handle, data->event_type});
+}
+
 extern "C" rmw_ret_t rmw_subscription_set_listener_callback(
   rmw_subscription_t * rmw_subscription,
   rmw_listener_callback_t callback,
   const void * user_data,
   const void * subscription_handle)
 {
-  (void)rmw_subscription;
-  (void)callback;
-  (void)user_data;
-  (void)subscription_handle;
-  // auto subscription = static_cast<CddsSubscription *>(rmw_subscription->data);
-  // subscription->setCallback(user_data, callback, subscription_handle);
-  RCUTILS_LOG_ERROR_NAMED(
-    "rmw_node.cpp",
-    "rmw_subscription_set_listener_callback: not supported (yet)");
-  return RMW_RET_UNSUPPORTED;
+  auto sub = static_cast<CddsSubscription *>(rmw_subscription->data);
+  auto entity_to_listen = sub->enth;
+
+  // This API sets two callbacks:
+  // 1. The DDS subscription's listener callback.
+  // 2. The user callback, called from the listener callback.
+  /*
+      dds_listener_callback(..) {
```

Comment:
I find this comment confusing.
this API only set a single thing which is an entity callback provided by the user.
what you describe is how cyclone does that, don't think is relevant here

Reply:
removed in 010fb92